# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/MediaEngineWebRTCAudio.cpp
# Commit: 0ae30eae800d
# Full Hash: 0ae30eae800dcad73d997f9f12bb56eb537ef86c
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2020-11-06 04:14:34
# Regressor Bug: 1651745
# File Overlap Count: 1
# Description:
#   Bug 1651745 - Don't buffer more than necessary in AudioInputProcessing. r=padenot
#   
#   When opening a second input track, there will already be some data from its
#   first instantiation in the driver's scratch buffer. If we ignore this data, we
#   end up buffering too much in AudioInputProcessing::Pull and tripping an assert.
# ==============================================================================

diff -r a00ad7266f54 -r 0ae30eae800d dom/media/webrtc/MediaEngineWebRTCAudio.cpp
--- a/dom/media/webrtc/MediaEngineWebRTCAudio.cpp	Thu Nov 05 16:43:24 2020 +0000
+++ b/dom/media/webrtc/MediaEngineWebRTCAudio.cpp	Thu Nov 05 16:44:00 2020 +0000
@@ -1133,12 +1133,19 @@
 void AudioInputProcessing::NotifyInputData(MediaTrackGraphImpl* aGraph,
                                            const AudioDataValue* aBuffer,
                                            size_t aFrames, TrackRate aRate,
-                                           uint32_t aChannels) {
+                                           uint32_t aChannels,
+                                           uint32_t aAlreadyBuffered) {
   MOZ_ASSERT(aGraph->OnGraphThread());
   TRACE();
 
   MOZ_ASSERT(mEnabled);
 
+  if (!mLiveFramesAppended) {
+    // First time we see live frames getting added. Use what's already buffered
+    // in the driver's scratch buffer as a starting point.
+    mLiveBufferingAppended = aAlreadyBuffered;
+  }
+
   // If some processing is necessary, packetize and insert in the WebRTC.org
   // code. Otherwise, directly insert the mic data in the MTG, bypassing all
   // processing.